pennylane/transforms/decompositions/two_qubit_unitary.py

```diff
@@ -228,62 +221,48 @@ def _extract_su2su2_prefactors(U, V):
-        :align: center
-        :width: 100%
-        :target: javascript:void(0);
 
-    where :math:`A, B, C, D` are :math:`SU(2)` gates.
+def _decomposition_3_cnots(U, wires):
+    r"""The most general form of this decomposition is U = (A \otimes B) V (C \otimes D),
+    where V is as depicted in the circuit below:
+     -U- = -C--X--RZ(d)--C---------X--A-
+     -U- = -D--C--RY(b)--X--RY(a)--C--B-
+    """
 
-    Args:
-        U (tensor): A 4 x 4 unitary matrix.
-        wires (Union[Wires, Sequence[int] or int]): The wires on which to apply the operation.
+    # First we add a SWAP as per v1 of 0308033, which helps with some
+    # rearranging of gates in the decomposition (it will cancel out the fact
+    # that we need to add a SWAP to fix the determinant in another part later).
+    swap_U = np.exp(1j * np.pi / 4) * math.dot(math.cast_like(SWAP, U), U)
 
-    Returns:
-        list[qml.Operation]: A list of operations that represent the decomposition
-        of the matrix U.
-    """
-    # First, test if we have a tensor product of two single-qubit operations. If
-    # so, we don't actually need to do a decomposition. To test this, we can
-    # check if Edag U E is in SO(4) because of the isomorphism between SO(4) and
-    # SU(2) x SU(2).
-    test_so4 = qml.math.linalg.multi_dot([Edag, U, E])
-    if qml.math.isclose(qml.math.linalg.det(test_so4), 1.0) and qml.math.allclose(
-        qml.math.dot(test_so4, qml.math.T(test_so4)), qml.math.eye(4)
-    ):
-        A, B = _su2su2_to_tensor_products(U)
-        A_ops = zyz_decomposition(A, wires[0])
-        B_ops = zyz_decomposition(B, wires[1])
-        return A_ops + B_ops
-
-    # The final form of this decomposition is U = (A \otimes B) V (C \otimes D),
-    # as expressed in the circuit below.
-    # -U- = -C--X--RZ(d)--C---------X--A-|
-    # -U- = -D--C--RY(b)--X--RY(a)--C--B-|
+    # Choose the rotation angles of RZ, RY in the two-qubit decomposition.
+    # They are chosen as per Proposition V.1 in quant-ph/0308033 and are based
+    # on the phases of the eigenvalues of :math:`E^\dagger \gamma(U) E`, where
+    #    \gamma(U) = (E^\dag U E) (E^\dag U E)^T.
+    # The rotation angles can be computed as follows (any three eigenvalues can be used)
+    u = math.dot(Edag, math.dot(swap_U, E))
+    gammaU = math.dot(u, math.T(u))
+    evs, _ = math.linalg.eig(gammaU)
 
-    # First, we note that this method works only for SU(4) gates, meaning that
-    # we need to rescale the matrix by its determinant. Furthermore, we add a
-    # SWAP as per v1 of 0308033, which helps with some rearranging of gates in
-    # the decomposition (it will cancel out the fact that we need to add a SWAP
-    # to fix the determinant in another part later).
-    swap_U = qml.math.exp(1j * np.pi / 4) * qml.math.dot(SWAP, _convert_to_su4(U))
+    # To get consistent results in all interfaces, sort the eigenvalues first
+    angles = qml.math.sort([math.angle(ev) for ev in evs])
```

Comment:
would
```python
angles = math.sort(math.angle(evs))
```
work out of the box due to vectorization?

Reply:
I thought it would, burns out it does not... in the torch interface, `sort` returns the indices in addition to the sorted elements so it's two values instead of one, and then the indexing in the next line fails.

```python
>>> qml.math.sort(qml.math.angle(evs))
torch.return_types.sort(
values=tensor([-1.8967, -0.8074,  1.5122,  2.6933], dtype=torch.float64),
indices=tensor([1, 0, 3, 2]))
```

Since it's only 4 elements to sort, I'm going to leave as-is rather than writing a special case.

Reply:
> in the torch interface, sort returns the indices in addition to the sorted elements so it's two values instead of one, and then the indexing in the next line fails.

🤦 

> Since it's only 4 elements to sort, I'm going to leave as-is rather than writing a special case.

Makes sense!